plutip-server/Api/Handlers.hs

```diff
@@ -0,0 +1,108 @@
+import Test.Plutip.Internal.LocalCluster (startCluster, stopCluster)
+import Types
+  ( StartClusterRequest(StartClusterRequest, keysToGenerate)
+  , StopClusterRequest(StopClusterRequest)
+  , StopClusterResponse(StopClusterSuccess, StopClusterFailure)
+  , AppM
+  , PrivateKey
+  , StartClusterResponse
+    ( ClusterStartupSuccess
+    , keysDirectory
+    , nodeSocketPath
+    , privateKeys
+    , nodeConfigPath
+    , ClusterStartupFailure
+    )
+  , ServerOptions(ServerOptions, nodeLogs)
+  , Env(status)
+  , Lovelace(unLovelace)
+  , ClusterStartupFailureReason
+    ( ClusterIsRunningAlready
+    , NegativeLovelaces
+    , NodeConfigNotFound
+    )
+  )
+import Cardano.Api (serialiseToCBOR)
+import Cardano.Launcher.Node (nodeSocketFile)
+import Cardano.Wallet.Shelley.Launch.Cluster (RunningNode(RunningNode))
+import Control.Concurrent.MVar (takeMVar, isEmptyMVar, putMVar)
+import Control.Monad (unless)
+import Control.Monad.Except (runExceptT, throwError)
+import Control.Monad.Extra (unlessM)
+import Control.Monad.IO.Class (liftIO)
+import Data.ByteString.Base16 qualified as Base16
+import Data.Default (def)
+import Data.Foldable (for_)
+import Data.Text.Encoding qualified as Text
+import Data.Traversable (for)
+import System.Directory (doesFileExist)
+import System.FilePath (replaceFileName)
+import Test.Plutip.Config (relayNodeLogs, chainIndexPort)
+import Test.Plutip.Internal.BotPlutusInterface.Setup (keysDir)
+import Test.Plutip.Internal.BotPlutusInterface.Wallet (BpiWallet(signKey))
+import Test.Plutip.Internal.BotPlutusInterface.Wallet (addSomeWallet)
+import Test.Plutip.Internal.Types (ClusterEnv(runningNode))
+
+startClusterHandler :: ServerOptions -> StartClusterRequest -> AppM StartClusterResponse
+startClusterHandler
+  ServerOptions { nodeLogs }
+  StartClusterRequest { keysToGenerate } = interpret $ do
+  -- Check that lovelace amounts are positive
+  for_ keysToGenerate $ \lovelaceAmounts -> do
+    for_ lovelaceAmounts $ \lovelaces -> do
+      unless (unLovelace lovelaces > 0) $ do
+        throwError NegativeLovelaces
+  statusMVar <- asks status
+  isClusterDown <- liftIO $ isEmptyMVar statusMVar
+  unless isClusterDown $ throwError ClusterIsRunningAlready
+  let
+    cfg = def
+      { relayNodeLogs = nodeLogs
+      , chainIndexPort = Nothing
+      }
+  (statusTVar, res@(clusterEnv, _)) <- liftIO $ startCluster cfg setup
+  liftIO $ putMVar statusMVar statusTVar
+  let nodeConfigPath = getNodeConfigFile clusterEnv
+  -- safeguard against directory tree structure changes
+  unlessM (liftIO $ doesFileExist nodeConfigPath) $ throwError NodeConfigNotFound
+  pure $ ClusterStartupSuccess
+    { privateKeys = getWalletPrivateKey <$> snd res
+    , nodeSocketPath = getNodeSocketFile clusterEnv
+    , nodeConfigPath = nodeConfigPath
+    , keysDirectory = keysDir clusterEnv
+    }
+  where
+    setup :: ReaderT ClusterEnv IO (ClusterEnv, [BpiWallet])
+    setup = do
+      env <- ask
+      wallets <- do
```

Comment:
Minor stylistic thing, but the `do` is redundant

Reply:
I prefer to leave `do` to make it clear that the context is monadic and to make diffs less noisy in case new lines are added.